packages/contracts/src/utils/asset_proxy_utils.ts

```diff
@@ -8,25 +8,62 @@ export const assetProxyUtils = {
     encodeAssetProxyId(assetProxyId: AssetProxyId): Buffer {
         return ethUtil.toBuffer(assetProxyId);
     },
+    decodeAssetProxyId(encodedAssetProxyId: Buffer): AssetProxyId {
+        return ethUtil.bufferToInt(encodedAssetProxyId);
+    },
     encodeAddress(address: string): Buffer {
         if (!ethUtil.isValidAddress(address)) {
             throw new Error(`Invalid Address: ${address}`);
         }
         const encodedAddress = ethUtil.toBuffer(address);
         return encodedAddress;
     },
+    decodeAddress(encodedAddress: Buffer): string {
+        const address = ethUtil.bufferToHex(encodedAddress);
+        if (!ethUtil.isValidAddress(address)) {
+            throw new Error(`Invalid Address: ${address}`);
+        }
+        return address;
+    },
     encodeUint256(value: BigNumber): Buffer {
         const formattedValue = new BN(value.toString(10));
         const encodedValue = ethUtil.toBuffer(formattedValue);
         return encodedValue;
     },
+    decodeUint256(encodedValue: Buffer): BigNumber {
+        const formattedValue = ethUtil.bufferToHex(encodedValue);
+        const value = new BigNumber(formattedValue, 16);
+        return value;
+    },
     encodeERC20ProxyData(tokenAddress: string): string {
         const encodedAssetProxyId = assetProxyUtils.encodeAssetProxyId(AssetProxyId.ERC20);
         const encodedAddress = assetProxyUtils.encodeAddress(tokenAddress);
         const encodedMetadata = Buffer.concat([encodedAssetProxyId, encodedAddress]);
         const encodedMetadataHex = ethUtil.bufferToHex(encodedMetadata);
         return encodedMetadataHex;
     },
+    decodeERC20ProxyData(proxyData: string): string /* tokenAddress */ {
```

Comment:
I would like if this had a type which includes the asset proxy Id. And also include a `decodeProxyData` which returns id, token address (optional) and data (optional). Forwarding wrapper currently switches on the asset id and has to pull it out.

Reply:
I've updated `decodeERC20ProxyData` & `decodeERC721ProxyData` accordingly. Also see the newly created function `decodeProxyData`, which uses the generalized form you requested.